multi/multi.go

```diff
@@ -0,0 +1,271 @@
+package multi
+
+import (
+	"errors"
+	"github.com/tarantool/go-tarantool"
+	"log"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	connDisconnected = iota
+	connConnected
+	connClosed
+)
+
+type ConnectionMulti struct {
+	addrs    []string
+	connOpts tarantool.Opts
+	opts     OptsMulti
+
+	mutex           sync.RWMutex
+	notify			chan tarantool.ConnEvent
+	state           uint32
+	control			chan struct{}
+	currentPoolAddr string
+	pool            map[string]*tarantool.Connection
+}
+
+var _ = tarantool.Connector(&ConnectionMulti{}) // check compatibility with connector interface
+
+type OptsMulti struct {
+	Chooser ConnectionChooser
+}
+
+func ConnectWithOpts(addrs []string, connOpts tarantool.Opts, opts OptsMulti) (connMulti *ConnectionMulti, err error) {
+	if len(addrs) == 0 {
+		return nil, errors.New("addrs should not be empty")
+	}
+	if opts.Chooser == nil {
+		return nil, errors.New("OptsMulti.Chooser should not be nil")
+	}
+
+	notify := make(chan tarantool.ConnEvent)
+	connOpts.Notify = notify
+	connMulti = &ConnectionMulti{
+		addrs:    addrs,
+		connOpts: connOpts,
+		opts:     opts,
+		notify:	  notify,
+		control:  make(chan struct{}),
+		pool:     make(map[string]*tarantool.Connection),
+	}
+
+	if _, err := connMulti.getNextConnection(false); err != nil {
+		return nil, err
+	}
+
+	go connMulti.checker()
+
+	return connMulti, nil
+}
+
+func Connect(addrs []string, connOpts tarantool.Opts) (connMulti *ConnectionMulti, err error) {
+	opts := OptsMulti{
+		Chooser: &RoundRobinChooser{},
+	}
+	return ConnectWithOpts(addrs, connOpts, opts)
+}
+
+func (connMulti *ConnectionMulti) checker() {
+	for connMulti.state != connClosed {
+		current := connMulti.getCurrentConnection()
+		select {
+		case <-connMulti.control:
+			return
+		case e := <-connMulti.notify:
+			if atomic.LoadUint32(&connMulti.state) != connClosed && current.Addr() == e.Conn.Addr() {
+				log.Printf("tarantoolmulti: disconnected from %s, trying another one", current.Addr())
+				if e.Kind == tarantool.Closed {
+					atomic.StoreUint32(&connMulti.state, connDisconnected)
+					connMulti.getNextConnection(true)
+				}
+				if e.Kind == tarantool.Disconnected {
+					atomic.StoreUint32(&connMulti.state, connDisconnected)
+				}
+			}
+		}
+	}
+}
+
+func (connMulti *ConnectionMulti) getNextConnection(reconnect bool) (*tarantool.Connection, error) {
+	var nextConn *tarantool.Connection
+	var err error
+
+	connMulti.mutex.Lock()
+	defer connMulti.mutex.Unlock()
+
+	for {
+		addr, shouldContinue := connMulti.opts.Chooser.GetNext(connMulti.addrs, reconnect)
+		if !shouldContinue {
+			break
+		}
+		nextConn, err = tarantool.Connect(addr, connMulti.connOpts)
+		if err == nil {
+			connMulti.currentPoolAddr = addr
+			connMulti.pool[addr] = nextConn
+			connMulti.state = connConnected
```

Comment:
value is written without atomic, although it is read then by atomic.

You'd better add access methods to it then.